### PR TITLE
Improved muon track validation (step1: base code deployment, no change to workflows)

### PR DESCRIPTION
**Pull request by abbiendi**
This is the first step in deployment of an improved muon track validation code.
The strategy is agreed with the Muon POG and PPD conveners.

This PR contains new code and configuration files, introduced alongside with the current code and cfgs. The new code is not yet used in the standard workflows, i.e. nothing should change anywhere.

The second step will be the customization of a few configuration files (marked by a comment) to let the new and the default code to run (both) for a given 81X prerelease cycle in few selected muon RelVals. This will allow for a smooth transition between the current and the new validation code. The plan foresees to deploy the new validation at the prerelease following the one with duplicate validations. At that point the old code should be possibly removed.
As I will be on vacation for the next few weeks this is supposed to let others to implement the necessary customization to try to speed up things a bit.
Also this base code is convenient for other developments (new code for upgrades, which can be integrated afterwards).

As a side effect the main muon POG validation affects the HIN validation of muons. HIN should decide what to do during this intermediate period. (I am available to help out, I tried to update the HIN part as well).

The new code and configuration has been flagged by a prefix "New" in filenames, class/module names, when it is a duplicate of the current production code. 
Features:
- much clearer code, easier maintenance and developments, some technical improvements
- new plots for performance vs PU, charge misid ...
- removal of a few unused plots
- improved settings of the histogram binning/ranges for different types of muons. All such settings contained in a cfg file "histoParameters_cff.py".
- better choice of names for histograms, corrected labels etc...
- better use of selectors to clean up the cfgs
- slimmed/clearer cfgs as a consequence of the above changes
- kept legacy configurations for use of TrackAssociatorByPosition in separate files, avoiding to clutter the main cfgs as in the current version
- new default setting of the MuonAssociatorByHits to flag as fakes global muons with tracker and muon hits matched to different particles: this will produce sensible fake rates (much higher than what is obtained currently). Side note: the current configuration has always lived for historical reasons.

cheers
    Giovanni

### CODE REVIEW

**Comment by slava77** [comment]
PR title should be updated to be more descriptive than "Dev new validation 81 x".
There should also be some description including a link to a presentation in a meeting.

**Comment by abbiendi** [comment]
Hi, sorry I do not know why it was sent right away, before I wrote anything... Can I change the title now ? I'm writing the comment.

**Comment by slava77** [comment]
On 7/8/16 10:19 AM, abbiendi wrote:
> Hi, sorry I do not know why it was sent right away, before I wrote
> anything... Can I change the title now ? I'm writing the comment.

There is an "Edit" button to edit the title and a pencil-like icon in
the description blob to edit that as well

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by smuzaffar** [comment]
@abbiendi , sorry, in order to get the IB pages recognize latest IBs I had to amend the e495908 commit to have message like "Merge CMSSW_8_0_X into CMSSW_8_1_X.". Due to that now to old commit is shows as part of this PR. Either ignore it (it is already part of IB) or please rebase based on CMSSW_8_1_X head. you can start from CMSSW_8_1_X and cherry pick the commits.

Just follow the instructions (replace your github user, branch and commits) https://github.com/cms-sw/cmssw/pull/15149#issuecomment-231485965

**Comment by abbiendi** [comment]
Hi, for me this is no problem. Fine if I can ignore it, since I am in vacation and do not have a decent connection...
(I do not know why my PR included automatically yours, it was like magic for me, but I assumed it was not a problem to go ahead)

Giovanni

On Jul 11, 2016, at 9:44 AM, Malik Shahzad Muzaffar wrote:


@abbiendi<https://github.com/abbiendi> , sorry, in order to get the IB pages recognize latest IBs I had to amend the e495908<https://github.com/cms-sw/cmssw/commit/e495908ec8dc510772f0b34865d530bee39cf4cd> commit to have message like "Merge CMSSW_8_0_X into CMSSW_8_1_X.". Due to that now to old commit is shows as part of this PR. Either ignore it (it is already part of IB) or please rebase based on CMSSW_8_1_X head. you can start from CMSSW_8_1_X and cherry pick the commits.

Just follow the instructions (replace your github user, branch and commits) #15149 (comment)<https://github.com/cms-sw/cmssw/pull/15149#issuecomment-231485965>

**Comment by slava77** [comment]
The code changes are not really related to reco algorithms.
My comments below are from rather general code evolution perspective.

It looks like the change pattern is to copy existing code with "New" prefix in file names and rather minor modifications, this is done together with many commented out pieces of the "New" in the current default configurations.
Given that this is not a complete rewrite, it would be more appropriate to copy the current stuff to "Old" and modify existing code to preserve development history.

Is there machinery in place to make comparisons between different validation modules (the old and new version), when both old and new modules are running?

If the idea is to have a switch-over step where both versions of validation exist (one to be used for comparisons with older release, and the other to be used for comparisons with future releases),
then this approach can only succeed if the pre-release is expected to have no issues,
just a stepping stone. I don't think this will work given the current development plans on the physics side and a need to check run2, 2017, and 2023 workflows in different degree of completion. We'll end up running both versions at least until the last bugfix version of 81X release.

**Comment by slava77** [comment]
ping
I was wondering if some activity is expected in this PR soon.

**Comment by abbiendi** [comment]
Hi Slava,
as said in my first description, the strategy has been suggested by POG and PPD conveners, so I'm waiting for helpful answers from them here.

**Comment by slava77** [comment]
Hi Carlo

Thank you for clarifying the strategy in the follow up to my comments.
I think it is reasonable to proceed with it.

I will not comment on code specifics, this is in DQM hands.
@dmitrijus your review and signature are needed here.

**Comment by slava77** [comment]
@abbiendi 
this PR doesn't merge anymore
Please rebase or update by other means

**Comment by abbiendi** [comment]
Hi Slava, 
I have updated it, there were just one trivial thing to edit in:
Validation/Configuration/python/postValidation_cff.py
and another one in:
DQM/TrackingMonitorSource/python/TrackingSourceConfig_Tier0_cff.py  (don't know why!)

I followed the instructions at: http://cms-sw.github.io/tutorial-forward-port.html
I do not understand why the PR apparently inherited a lot of other changed files.
(maybe just because I was waiting for pending signatures since long)

   Giovanni

**Comment by slava77** [comment]
On 8/26/16 9:33 AM, abbiendi wrote:
> Hi Slava,
> I have updated it, there were just one trivial thing to edit in:
> Validation/Configuration/python/postValidation_cff.py
> and another one in:
> DQM/TrackingMonitorSource/python/TrackingSourceConfig_Tier0_cff.py
> (don't know why!)
> 
> I followed the instructions at:
> http://cms-sw.github.io/tutorial-forward-port.html
> I do not understand why the PR apparently inherited a lot of other
> changed files.
> (maybe just because I was waiting for pending signatures since long)
> 
> Giovanni
> 

Hi Giovanni,

in this tutorial you should follow the rebase instructions.
(search for "rebase")


		--slava

**Comment by mmusich** [comment]
@abbiendi, hi Giovanni, this PR needs a rebase

**Comment by abbiendi** [comment]
I never did a rebase until now...I tried and it went smooth until before the final "git push". At that point I got an error message which I did not understand. However if I look at the PR now it seems the history has been cleaned. Since I was not sure I have then pushed to a new branch called "devNewValidation_81X_rebased". Please suggest what to do.
thanks,    Giovanni

**Comment by slava77** [comment]
@abbiendi 
once your rebase succeeded on the local side, 
you should ``` git push --force ``` to the remote and you can use the same branch to update this PR automatically

**Comment by abbiendi** [comment]
sorry if I'm making a lot of noise... 
What I did now is:

[giovanni@lxplus126 src]$ git push my-cmssw HEAD:devNewValidation_81X
Enter passphrase for key '/afs/cern.ch/user/g/giovanni/.ssh/id_rsa': 
To git@github.com:abbiendi/cmssw.git
 ! [rejected]        HEAD -> devNewValidation_81X (non-fast-forward)
error: failed to push some refs to 'git@github.com:abbiendi/cmssw.git'
hint: Updates were rejected because a pushed branch tip is behind its remote
hint: counterpart. Check out this branch and merge the remote changes
hint: (e.g. 'git pull') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
[giovanni@lxplus126 src]$ git push --force my-cmssw HEAD:devNewValidation_81X
Enter passphrase for key '/afs/cern.ch/user/g/giovanni/.ssh/id_rsa': 
Counting objects: 142, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (47/47), done.
Writing objects: 100% (115/115), 48.80 KiB | 0 bytes/s, done.
Total 115 (delta 79), reused 87 (delta 68)
remote: Resolving deltas: 100% (79/79), completed with 22 local objects.
To git@github.com:abbiendi/cmssw.git
 + 119339d...1ce47da HEAD -> devNewValidation_81X (forced update)

since I still do not understand the meaning of the trouble, I cross my fingers, hope it is all fine this time

 thanks,   Giovanni

**Comment by mmusich** [comment]
@abbiendi (at least the history) looks good

**Comment by mmusich** [comment]
please test

**Comment by slava77** [comment]
ping
I was going to wait for the primary L2 area (DQM) to review this before signing off.
Please let me know the status.
Thanks.

**Comment by abbiendi** [comment]
thanks Slava: nothing is needed on my side, I was also waiting for others' signoff

Giovanni

On Sep 12, 2016, at 11:18 PM, Slava Krutelyov wrote:


ping
I was going to wait for the primary L2 area (DQM) to review this before signing off.
Please let me know the status.
Thanks.

**Review by dmitrijus — APPROVED** [review]

**Comment by dmitrijus** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/15153 1ce47da
- validation code changes; reco is minimally affected
- jenkins tests passed
- local checks show that expanded configs for reco step (step3) in 2016 data and MC workflows are the same.

File: Configuration/StandardSequences/python/ValidationCosmics_cff.py
Comment (franzoni) [inline Configuration/StandardSequences/python/ValidationCosmics_cff.py]: hello @abbiendi @HuguesBrun 
what is the reason why you rename these two sequences,
given that you don't use them anywhere else in your PR?

I've checked in various test.py scattered are around 81X (outside the central sequences exercised by the short matrix)
and nowhere could I find where the prevalidation which is defined in ValidationCosmics_cff.py  is used

As such, this PR could go... if I haven't missed anything!

Comment (abbiendi) [inline Configuration/StandardSequences/python/ValidationCosmics_cff.py]: Hi Giovanni,

if I remember, this was just to follow a similar nomenclature as in other configuration files.

**Comment by slava77** [comment]
@abbiendi @davidlange6 
I'm curious if this is going anywhere

**Comment by abbiendi** [comment]
I guess the question here is for David, I have been waiting quite long for (few) others to approve...

(as a safety measurement including @battibass @HuguesBrun as well)

File: Configuration/StandardSequences/python/ValidationCosmics_cff.py
Comment (battibass) [inline Configuration/StandardSequences/python/ValidationCosmics_cff.py]: @franzoni , apologies for replying with a long delay on this.
This part of the validation machinery was not excercised in cosmics in a central workflow (was just used in workflows run privately, at least in recent past).
Anyhow MUO has people who joined the team in the past months that, after working on data certification for cosmics for 2016, plan to improve matrix workflows as 1307 .
That sequence (not included at present) is a good hook for them to start improving the monitoring in the workflow mentioned above.
(I've just discussed with @HuguesBrun that this is still in MUO plans)

**Comment by battibass** [comment]
As we have been mentioned in the last reply from @abbiendi:
@HuguesBrun confirms that MUO is still interested in pushing for the inclusion of the PR as it is.
Moreover, if included in a close-to-stable period of the release development cycles (i.e. soon), this would significantly simplify the tests for the switch-over step mentioned earlier in this thread.

**Comment by slava77** [comment]
-1

- no more developments in 81X. 
this 81X PR should be closed.

**Comment by franzoni** [comment]
hello @abbiendi 
is this PR still actual ?
If not, would it be possible to close it?
Thanks. g.

**Comment by abbiendi** [comment]
Hi, yes this should be closed as #16646 has been closed (they were based on the same branch). I opened a new PR in 90X, #17489 , that has the same purpose and should be hopefully merged.

**Comment by franzoni** [comment]
hello @abbiendi 
is this PR still of interest ?
if not, maybe it's worthwhile closing it ?

**Comment by abbiendi** [comment]
Hello @franzoni : why are you interested in this old PR ?  I already answered 4 days ago ! 
I am not an offline manager, I can't close anything...
instead I am waiting to see my newer PR #17489 approved, but for the same reason I can't do anything than waiting... please ping the new PR thread in case...
thanks

**Comment by boudoul** [comment]
Hello @abbiendi , for this one, just push the 'close pull request ' button of this PR , which can done by the original author of the PR

**Comment by abbiendi** [comment]
ah ok, never thought of it before !

**Pull request closed without merging**
